### PR TITLE
prevent infinite loop if there are more duplicate property_key than batch_size

### PR DESCRIPTION
**Pull request by ericcj**
by excluding seen ids

### CODE REVIEW

**Comment by ericcj** [comment]
@nambrot this infinite loop is a pretty rough edge, but this does deprecate support for strict_mode which is incompatible with it

**Comment by nambrot** [comment]
Could you elaborate/document a little more the edge case you are referring to allow other users to understand the change as well as the need for deprecation of strict mode

**Comment by ericcj** [comment]
If there are duplicates in the key beyond ordered by, this gem will loop infinitely trying to fetch the next batch.

**Comment by nambrot** [comment]
It's been a while since I've been using this module, but as far as I can tell, it does not invalidate the original use case as well as strict mode for existing users. Thus I would be reluctant to change the semantics of the gem so drastically.

However, your edge case of the cardinality of a key being larger than the batch size being a valid one as well. I'm not sure though if there is a valid automatic strategy that we can prescribe as we potentially skip records, no?

**Comment by ericcj** [comment]
I could make this uniq_mode optional

**Comment by nambrot** [comment]
That would certainly be preferable. I do think a general "id-exclusion" logic could be useful in general, not only for this specific edge case

**Comment by ericcj** [comment]
I mis-remembered this actually.  strict_mode means "return records with same values as the last record of the last batch," e.g. use <= or >= inclusive, which was the default and I maintained here, except I fixed the bug where too many records with the same value as the last batch would cause an infinite loop by explicitly excluding those with the same id.  

I deprecated strict_mode=false which used exclusive < or > operators to exclude records with the same value as the last record of the last batch because it never seemed desirable to silently skip over records if we can safely iterate through them by always using inclusive operators and explicitly rejecting records with the same id as I've done here.

**Comment by nambrot** [comment]
Do you know the performance implications of excluding records by ID? I'm not sure we should do this by default for everyone unless they expect to run into this issue a lot (which in the use cases I'm aware of is very unlikely).

I think strict_mode=false has a valid use case as well as if one has a unique property key in which case you get to avoid deduping records as well as maintaining the start_ids logic.

I definitely agree that having this built in here is useful, I would like to maintain backwards compatibility however as there are valid reasons for strict_more=true/false

**Comment by ericcj** [comment]
I agree I'd like to maintain backward compatibility.  I started doing so for strict_mode=false, but quickly realized that strict_mode=true is not just the default but actually the only mode since it is parsed as "options.delete(:strict_mode) || true" which always evaluates to true even if you pass it strict_mode: false.  given that, the only thing I could find that wasn't 100% backward compatible was querying by "id" so i fixed it to match the logic in active record eql? and respect different primary_key columns.  

in the typical case of a unique order column we send a single "WHERE id != 1234" clause only for the last value of the page, which should have very low performance impact.  in the cases where there are duplicate values, excluding them from the query almost certainly has better performance than returning them each time and skipping past them in ruby.  it certainly has a better failure mode of sending a large exclusion clause than "itereate perpetually" querying your database in an infinite loop.  i'd guess that common uses of this gem are to iterate by timestamps, in which case duplicates are more common than we might expect because of special values like epoch time, distantPast, etc.

File: lib/ar-find-in-batches-with-order.rb [hunk 1]
```diff
@@ -30,27 +27,27 @@ def find_in_batches_with_order(options = {})
 
         break if records_size < batch_size
 
-        start = records.last.try(property_key)
+        next_start = records.last.try(property_key)
+        with_start_ids.clear if start != next_start
+        start = next_start
 
-        records = strict_mode ? (direction == :desc ? relation.where("#{sanitized_key} <= ?", start).to_a : relation.where("#{sanitized_key} >= ?", start).to_a) : (direction == :desc ? relation.where("#{sanitized_key} < ?", start).to_a : relation.where("#{sanitized_key} > ?", start).to_a)
+        records.each do |record|
```
Comment (nambrot) [inline lib/ar-find-in-batches-with-order.rb:35]: What you think of a more functional approach a la 

```
with_start_ids = records.filter { |record| record.try(property_key) == start }.map(&:id)
```

File: lib/ar-find-in-batches-with-order.rb [hunk 2]
```diff
@@ -30,27 +27,27 @@ def find_in_batches_with_order(options = {})
...
+          if record.try(property_key) == start
+            with_start_ids << record.id
+          end
+        end
+
+        without_dups = relation.where.not(relation.klass.primary_key => with_start_ids)
```
Comment (nambrot) [inline lib/ar-find-in-batches-with-order.rb:41]: Above you assume `.id` for the records, but here you precisely extract the actual primary key column, so maybe apply the same above?

Comment (nambrot) [reply lib/ar-find-in-batches-with-order.rb:41]: Should we avoid even adding the `.not` if `start_ids` is empty?

File: lib/ar-find-in-batches-with-order.rb (hunk 1 above)
Comment (ericcj) [reply lib/ar-find-in-batches-with-order.rb:35]: it needs to maintain the list of dups from the previous batch so we can't re-assign it here

File: lib/ar-find-in-batches-with-order.rb (hunk 2 above)
Comment (ericcj) [reply lib/ar-find-in-batches-with-order.rb:41]: the id method actually uses primary_key:  https://apidock.com/rails/ActiveRecord/Base/id

with_start_ids always includes at least the key from records.last so it is never empty

**Review by nambrot — APPROVED** [review]

**Comment by nambrot** [comment]
Is this good to merge @ericcj ?

**Comment by ericcj** [comment]
@nambrot oh yes!  running happily in production

**Comment by nambrot** [comment]
Thanks for your contribution!

**Pull request merged**